src/main/java/ch/njol/skript/structures/StructVariables.java

```diff
@@ -66,26 +78,92 @@ public class StructVariables extends Structure {
 		Skript.registerStructure(StructVariables.class, "variables");
 	}
 
-	private final List<NonNullPair<String, Object>> variables = new ArrayList<>();
+	public static class DefaultVariables implements ScriptData {
+
+		private final List<NonNullPair<String, Object>> variables = new ArrayList<>();
+		private final Deque<Map<String, Class<?>[]>> hints = new ArrayDeque<>();
+
+		public DefaultVariables(Collection<NonNullPair<String, Object>> variables) {
+			this.variables.addAll(variables);
+		}
+
+		@SuppressWarnings("unchecked")
+		public void add(String variable, Class<?>... hints) {
+			if (hints == null || hints.length <= 0)
+				return;
+			if (CollectionUtils.containsAll(hints, Object.class)) // Ignore useless type hint
+				return;
+			// Take top of stack, without removing it
+			Map<String, Class<?>[]> head = this.hints.getFirst();
+			head.put(variable, hints);
+		}
+
+		public void enterScope() {
+			hints.push(new HashMap<>());
+		}
+
+		public void exitScope() {
+			hints.pop();
+		}
+
+		/**
+		 * Returns the type hints of a variable.
+		 * Can be null if no type hint was saved.
+		 * 
+		 * @param variable The variable string of a variable.
+		 * @return type hints of a variable if found otherwise null.
+		 */
+		@Nullable
+		public Class<?>[] get(String variable) {
+			for (Map<String, Class<?>[]> map : hints) {
+				Class<?>[] hints = map.get(variable);
+				if (hints != null && hints.length > 0)
+					return hints;
+			}
+			return null;
+		}
+
+		public boolean hasDefaultVariables() {
+			return !variables.isEmpty();
+		}
+
+		/**
+		 * @return an unmodifiable list of all the default variables registered for the script.
+		 */
+		@Unmodifiable
+		public List<NonNullPair<String, Object>> getVariables() {
+			return Collections.unmodifiableList(variables);
+		}
+	}
 
 	@Override
 	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult, EntryContainer entryContainer) {
-		// TODO allow to make these override existing variables
 		SectionNode node = entryContainer.getSource();
 		node.convertToEntries(0, "=");
+
+		List<NonNullPair<String, Object>> variables = new ArrayList<>();
 		for (Node n : node) {
 			if (!(n instanceof EntryNode)) {
-				Skript.error("Invalid line in variables section");
+				Skript.error("Invalid line in variables structure");
 				continue;
 			}
 
 			String name = n.getKey().toLowerCase(Locale.ENGLISH);
 			if (name.startsWith("{") && name.endsWith("}"))
 				name = name.substring(1, name.length() - 1);
 
-			String var = name;
+			if (name.startsWith(Variable.LOCAL_VARIABLE_TOKEN)) {
+				Skript.error("'" + name + "' cannot be a local variable in default variables structure");
+				continue;
+			}
+
+			if (name.contains("<") || name.contains(">")) {
+				Skript.error("'" + name + "' cannot have symbol '<' or '>' within the definition");
+				continue;
+			}
```

Comment:
I know you added this in response to my previous comment in VariableString, but I'm not sure this is exactly what I meant. If this restriction isn't taken for regular variables, I don't think it's necessary here.

Reply:
Since it's a breaking change now, would we be better off putting this change in that PR?